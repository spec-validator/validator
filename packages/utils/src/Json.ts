export type Primitive =
| string
| number
| boolean

export type Nothing =
| null
| undefined

export type Json =
| Primitive
| Nothing
| { [property: string]: Json }
| readonly Json[]
