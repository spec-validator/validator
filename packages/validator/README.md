# @spec-validator/validator

## Core API

Fields can be defined by implementing a `Field` interface via a `declareField` function.

The first parameteter of `declareField` is a string value representing field type and the
second parameter is a function that aims to construct a field object.

`declareField` returns a function with the same interface as the constructor one passed
as the second parameter.

```ts
import assert from 'assert'

import { Field, declareField } from '@spec-validator/validator/core'

const DAYS = ['MO', 'TU', 'WE', 'TH', 'FR', 'SA', 'SU'] as const

const choices = new Set(DAYS)

type Day = typeof DAYS[number]

const dayField = declareField('DateField', (): Field<Day> => ({
  validate: (value: any): Day => {
    if (!choices.has(value)) {
      throw 'Invalid day'
    }
    return value
  },
  serialize: (deserialized) => deserialized,
}))
```

Define a schema:

```ts
const schema = {
  day: dayField(),
}
```

Validate and serialize payload:

```ts
import { validate, serialize } from '@spec-validator/validator'

const deserialized = validate(schema, { day: 'MO' })

assert.deepStrictEqual(
  deserialized,
  {
    day: 'MO'
  }
)

const serialized = serialize(schema, deserialized)

assert.deepStrictEqual(
  serialized,
  {
    day: 'MO'
  }
)
```

Errors are reported in a structured form with a `path` outlining
location of the error.

```ts
let error: any = undefined

try {
  validate(schema, {
    day: 'Abracadabra',
  })
} catch (err) {
  error = err
}

assert.deepStrictEqual(error, {
  inner: "Invalid day",
  path: [
    'day'
  ]
})
```

Infer TypeScript type from schema:

```ts
import { TypeHint } from '@spec-validator/validator'
import { expectType } from '@spec-validator/test-utils/expectType'

type Schema = TypeHint<typeof schema>

expectType<Schema, {
  day: Day
}>(true)
```

## Fields

### Primitive fields

```ts
import {
  booleanField, numberField, stringField
} from '@spec-validator/validator/fields'

const primitiveSpec = {
  flag: booleanField(),
  title: stringField(),
  // regex for email
  email: stringField(/\S+@\S+\.\S+/),
  count: numberField(),
  weight: numberField({ canBeFloat: true }),
  rating: numberField({ signed: true }),
}

type PrimitiveType = TypeHint<typeof primitiveSpec>

const valid: PrimitiveType = validate(primitiveSpec, {
  flag: true,
  title: 'Title',
  email: 'test@example.com',
  count: 12,
  weight: 16.33,
  rating: -10,
})

assert.deepStrictEqual(valid, {
  flag: true,
  title: 'Title',
  email: 'test@example.com',
  count: 12,
  weight: 16.33,
  rating: -10,
})
```

### Collection fields

Nested objects and arrays can be outlined as follows:

```ts
const complexSpec = {
  firstLevel: stringField(),
  innerObj: {
    secondLevel: stringField(),
    array: [{
      itemsThirdLevel: stringField()
    }]
  }
}

const validNestedObj = validate(complexSpec, {
  firstLevel: 'First level',
  innerObj: {
    secondLevel: 'Second level',
    array: [
      {
        itemsThirdLevel: 'Third level #1'
      }, {
        itemsThirdLevel: 'Third level #2'
      }
    ]
  }
})

assert.deepStrictEqual(!!validNestedObj, true)
```

### Auxilary fields

In addition to primitive and collection fields there are also ones that aim to
match more extensive type checking.

#### **optional**

Once annotated, a payload becomes `T | undefined`:

```ts
import { optional } from '@spec-validator/validator/fields'

const optionalField = optional(numberField())

expectType<TypeHint<typeof optionalField>, number | undefined>(true)

assert.deepStrictEqual(validate(optionalField, 42), 42)
assert.deepStrictEqual(validate(optionalField, undefined), undefined)
```

#### **choiceField**

A union type that may have only primitive types' based constants:

```ts
import { choiceField } from '@spec-validator/validator/fields'

const withChoices = choiceField(1, 2, 3)

expectType<TypeHint<typeof withChoices>, 1 | 2 | 3>(true)

assert.deepStrictEqual(validate(withChoices, 1), 1)
```

#### **dateField**

A mapping between JavaScript-friendly string date value and
a [Date](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date)
object.

```ts
import { dateField } from '@spec-validator/validator/fields'

const date = dateField()

expectType<TypeHint<typeof date>, Date>(true)

assert.deepStrictEqual(validate(date, '1995-12-17T03:24:00'), new Date('1995-12-17T03:24:00'))
```

#### **constantField**

Similar to `choiceField` but only with a single valid choice


```ts
import { constantField } from '@spec-validator/validator/fields'

const constant = constantField(11)

expectType<TypeHint<typeof constant>, 11>(true)

assert.deepStrictEqual(validate(constant, 11), 11)
```

#### **withDefault**

Similar to `optional` but instead of returning `undefined` in
case of a missing field.

```ts
import { withDefault } from '@spec-validator/validator/fields'

const defaultValue = withDefault(numberField(), 42)

expectType<TypeHint<typeof defaultValue>, number>(true)

assert.deepStrictEqual(validate(defaultValue, 11), 11)
assert.deepStrictEqual(validate(defaultValue, undefined), 42)
```

#### **unionField**

A field for the cases when an object can have multiple structures
(i.e. union of objects).

```ts
import { unionField } from '@spec-validator/validator/fields'

const union = unionField(
  {
    innerFieldV1: stringField(),
  },
  {
    innerFieldV2: numberField(),
  }
)

expectType<TypeHint<typeof union>, {
  innerFieldV1: string
} | {
  innerFieldV2: number
}>(true)

assert.deepStrictEqual(validate(union, {
  innerFieldV1: 'value'
}), {
  innerFieldV1: 'value'
})
assert.deepStrictEqual(validate(union, {
  innerFieldV2: 42
}), {
  innerFieldV2: 42
})
```

#### **wildcardObjectField**

In case if payload can be virtually anything JSON friendly (i.e. no validation is required)

```ts
import { wildcardObjectField } from '@spec-validator/validator/fields'

const wildCard = wildcardObjectField()

assert.deepStrictEqual(validate(wildCard, '"Abracadabra"'), 'Abracadabra')
```

### Segment field ($)

This special kind of a field with a fluent API, is for outlining a
structure of a string (e.g. a url pattern).

```ts
import { $ } from '@spec-validator/validator/fields'

const url = $
  ._('/')
  ._('username', stringField())
  ._('/items/')
  ._('id', numberField())

assert.deepStrictEqual(url.toString(), '^/(?<username>.*)/items/(?<id>\\d+)$')

expectType<TypeHint<typeof url>, {
  username: string,
  id: number
}>(true)

assert.deepStrictEqual(validate(url, '/john-smith/items/42'), {
  id: 42,
  username: 'john-smith'
})
```
