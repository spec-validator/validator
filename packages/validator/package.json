{
  "name": "@spec-validator/validator",
  "version": "0.0.1",
  "dependencies": {
    "@spec-validator/utils": "0.0.1"
  },
  "devDependencies": {
    "@spec-validator/test-utils": "0.0.1"
  }
}
