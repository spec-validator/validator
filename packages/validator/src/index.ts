export { SpecUnion, Field, TypeHint } from './core'
export { Json, Primitive } from './Json'

export { validate, serialize } from './interface'

import createRegistry from './registry'

export { createRegistry }
