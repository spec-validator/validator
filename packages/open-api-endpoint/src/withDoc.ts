import { declareField, Field } from '@spec-validator/validator/core'
import { Any } from '@spec-validator/utils/util-types'

type Example<T extends Any> = {
  value: T,
  summary: string
}

type Doc<T extends Any> = {
  description?: string,
  format?: string,
  examples?: Record<string, Example<T>>
}

export interface WithDoc<T extends Any, F extends Field<T>> extends Field<T> {
  readonly innerField: F,
  readonly doc: Doc<T>
}

export default declareField('@spec-validator/fields.WithDoc', <T extends Any, F extends Field<T>> (
  innerField: F,
  doc: Doc<T>
): WithDoc<T, F> => ({
    innerField,
    doc,
    validate: (it) => innerField.validate(it),
    serialize: (it) => innerField.serialize(it),
  }))
