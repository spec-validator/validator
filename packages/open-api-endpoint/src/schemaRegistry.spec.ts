import {
  arrayField, booleanField, choiceField, constantField,
  numberField, objectField, optional, stringField, unionField, wildcardObjectField, withDefault,
} from '@spec-validator/validator/fields'
import { Primitive } from '@spec-validator/validator/Json'
import getSchema from './schemaRegistry'
import withDoc from './withDoc'

test('BASE_PAIRS', () => {
  expect([
    arrayField(booleanField()),
    booleanField(),
    numberField(),
    numberField({canBeFloat: true}),
    wildcardObjectField(),
    choiceField(1, 2, 3),
    choiceField<Primitive>(1, true),
    choiceField(),
    constantField(42),
    objectField({
      a: booleanField(),
      b: booleanField(),
    }),
    objectField({
      opt: optional(stringField()),
    }),
    optional(booleanField()),
    stringField(),
    stringField(/\d+/),
    unionField(
      numberField(),
      stringField(),
      constantField(1),
      objectField({
        a: booleanField(),
        b: booleanField(),
      })
    ),
    withDefault(booleanField(), true),
    withDoc(booleanField(), {
      description: 'Answer',
      format: 'yes/no',
      examples: {
        'yes':{
          summary: 'positive',
          value: true,
        },
        'no': {
          summary: 'negative',
          value: false,
        }},
    }),
  ].map(field => getSchema(field))).toMatchSnapshot()
})
