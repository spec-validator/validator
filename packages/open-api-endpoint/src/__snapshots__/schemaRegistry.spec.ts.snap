// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BASE_PAIRS 1`] = `
Array [
  Object {
    "items": Object {
      "type": "boolean",
    },
    "type": "array",
  },
  Object {
    "type": "boolean",
  },
  Object {
    "minimum": 0,
    "type": "integer",
  },
  Object {
    "minimum": 0,
    "type": "number",
  },
  Object {
    "additionalProperties": true,
    "type": "object",
  },
  Object {
    "enum": Array [
      1,
      2,
      3,
    ],
    "type": "number",
  },
  Object {
    "oneOf": Array [
      Object {
        "enum": Array [
          1,
        ],
        "type": "number",
      },
      Object {
        "enum": Array [
          true,
        ],
        "type": "boolean",
      },
    ],
  },
  Object {},
  Object {
    "enum": Array [
      42,
    ],
    "type": "number",
  },
  Object {
    "additionalProperties": false,
    "properties": Object {
      "a": Object {
        "type": "boolean",
      },
      "b": Object {
        "type": "boolean",
      },
    },
    "required": Array [
      "a",
      "b",
    ],
    "type": "object",
  },
  Object {
    "additionalProperties": false,
    "properties": Object {
      "opt": Object {
        "pattern": ".*",
        "type": "string",
      },
    },
    "type": "object",
  },
  Object {
    "type": "boolean",
  },
  Object {
    "pattern": ".*",
    "type": "string",
  },
  Object {
    "pattern": "\\\\d+",
    "type": "string",
  },
  Object {
    "oneOf": Array [
      Object {
        "minimum": 0,
        "type": "integer",
      },
      Object {
        "pattern": ".*",
        "type": "string",
      },
      Object {
        "enum": Array [
          1,
        ],
        "type": "number",
      },
      Object {
        "additionalProperties": false,
        "properties": Object {
          "a": Object {
            "type": "boolean",
          },
          "b": Object {
            "type": "boolean",
          },
        },
        "required": Array [
          "a",
          "b",
        ],
        "type": "object",
      },
    ],
  },
  Object {
    "default": true,
    "type": "boolean",
  },
  Object {
    "description": Object {
      "description": "Answer",
      "examples": Object {
        "no": Object {
          "summary": "negative",
          "value": false,
        },
        "yes": Object {
          "summary": "positive",
          "value": true,
        },
      },
      "format": "yes/no",
    },
    "type": "boolean",
  },
]
`;
